File: frontend/src/api/project.ts
/**
 * Project API - 项目 API 调用模块
 */
import api from './index'
import type {
  ProjectCreate,
  ProjectUpdate,
  ProjectResponse,
  ProjectMember,
  ProjectMemberCreate,
  ProjectSubscriber,
  ProjectStatistics,
  ProjectIssuesSummary
} from '@/types/project'

// ==================== Project CRUD ====================

/**
 * 创建项目
 */
export async function createProject(
  workspaceId: number,
  data: ProjectCreate
): Promise<ProjectResponse> {
  let url = `/projects/?workspace_id=${workspaceId}`
  if (data.template_id) {
    url += `&template_id=${data.template_id}`
  }
  const response = await api.post(url, data)
  return response.data
}

/**
 * 列出工作空间的项目
 */
export async function listProjects(
  workspaceId: number,
  options?: {
    include_archived?: boolean
    limit?: number
    offset?: number
  }
): Promise<ProjectResponse[]> {
  const params = new URLSearchParams()
  params.append('workspace_id', workspaceId.toString())
  
  if (options?.include_archived) params.append('include_archived', 'true')
  if (options?.limit) params.append('limit', options.limit.toString())
  if (options?.offset) params.append('offset', options.offset.toString())
  
  const response = await api.get(`/projects/?${params.toString()}`)
  return response.data
}

/**
 * 获取项目详情
 */
export async function getProject(projectId: number): Promise<ProjectResponse> {
  const response = await api.get(`/projects/${projectId}`)
  return response.data
}

/**
 * 更新项目
 */
export async function updateProject(
  projectId: number,
  data: ProjectUpdate
): Promise<ProjectResponse> {
  const response = await api.patch(`/projects/${projectId}`, data)
  return response.data
}

/**
 * 删除项目
 */
export async function deleteProject(projectId: number): Promise<void> {
  await api.delete(`/projects/${projectId}`)
}

// ==================== Project Archive ====================

/**
 * 归档项目
 */
export async function archiveProject(projectId: number): Promise<ProjectResponse> {
  const response = await api.post(`/projects/${projectId}/archive`)
  return response.data
}

/**
 * 恢复项目
 */
export async function restoreProject(projectId: number): Promise<ProjectResponse> {
  const response = await api.post(`/projects/${projectId}/restore`)
  return response.data
}

// ==================== Project Members ====================

/**
 * 列出项目成员
 */
export async function listProjectMembers(
  projectId: number,
  onlyActive: boolean = true
): Promise<ProjectMember[]> {
  const response = await api.get(
    `/projects/${projectId}/members?only_active=${onlyActive}`
  )
  return response.data
}

/**
 * 添加项目成员
 */
export async function addProjectMember(
  projectId: number,
  data: ProjectMemberCreate
): Promise<{ id: number; project_id: number; user_id: number; role: number }> {
  const response = await api.post(
    `/projects/${projectId}/members?user_id=${data.user_id}&role=${data.role || 15}`
  )
  return response.data
}

/**
 * 更新项目成员角色
 */
export async function updateProjectMember(
  projectId: number,
  userId: number,
  role: number
): Promise<{ id: number; project_id: number; user_id: number; role: number }> {
  const response = await api.patch(
    `/projects/${projectId}/members/${userId}?role=${role}`
  )
  return response.data
}

/**
 * 移除项目成员
 */
export async function removeProjectMember(
  projectId: number,
  userId: number
): Promise<{ project_id: number; user_id: number; action: string }> {
  const response = await api.delete(`/projects/${projectId}/members/${userId}`)
  return response.data
}

// ==================== Project Lead & Default Assignee ====================

/**
 * 更新项目默认处理人
 */
export async function updateDefaultAssignee(
  projectId: number,
  userId: number | null
): Promise<ProjectResponse> {
  const response = await api.patch(
    `/projects/${projectId}?default_assignee_id=${userId ?? ''}`
  )
  return response.data
}

/**
 * 更新项目负责人
 */
export async function updateProjectLead(
  projectId: number,
  userId: number | null
): Promise<ProjectResponse> {
  const response = await api.patch(
    `/projects/${projectId}?project_lead_id=${userId ?? ''}`
  )
  return response.data
}

// ==================== Project Subscribers ====================

/**
 * 列出项目订阅者
 */
export async function listProjectSubscribers(
  projectId: number
): Promise<ProjectSubscriber[]> {
  const response = await api.get(`/projects/${projectId}/subscribers`)
  return response.data
}

/**
 * 添加项目订阅者
 */
export async function addProjectSubscriber(
  projectId: number,
  userId: number
): Promise<any> {
  const response = await api.post(
    `/projects/${projectId}/subscribers?user_id=${userId}`
  )
  return response.data
}

/**
 * 移除项目订阅者
 */
export async function removeProjectSubscriber(
  projectId: number,
  userId: number
): Promise<any> {
  const response = await api.delete(
    `/projects/${projectId}/subscribers/${userId}`
  )
  return response.data
}

// ==================== Project Statistics ====================

/**
 * 获取项目统计
 */
export async function getProjectStatistics(
  projectId: number
): Promise<ProjectStatistics> {
  const response = await api.get(`/projects/${projectId}/statistics`)
  return response.data
}

/**
 * 获取项目工作项摘要
 */
export async function getProjectIssuesSummary(
  projectId: number
): Promise<ProjectIssuesSummary> {
  const response = await api.get(`/projects/${projectId}/issues-summary`)
  return response.data
}

// ==================== Export all ====================

export const projectApi = {
  // CRUD
  createProject,
  listProjects,
  getProject,
  updateProject,
  deleteProject,
  
  // Archive
  archiveProject,
  restoreProject,
  
  // Members
  listProjectMembers,
  addProjectMember,
  updateProjectMember,
  removeProjectMember,
  
  // Lead & Default Assignee
  updateDefaultAssignee,
  updateProjectLead,
  
  // Subscribers
  listProjectSubscribers,
  addProjectSubscriber,
  removeProjectSubscriber,
  
  // Statistics
  getProjectStatistics,
  getProjectIssuesSummary
}

export default projectApi

File: backend/app/api/v1/endpoints/project.py
from fastapi import APIRouter, Depends, Path, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List

from app.db.session import get_db
from app.api.deps import get_current_user
from app.models.user import User
from app.schemas.project import ProjectCreate, ProjectUpdate, ProjectResponse
from app.services import project as project_service
from app.core.exceptions import NotFoundException, ConflictException

router = APIRouter()


# ==================== Project CRUD ====================

@router.post("/", response_model=ProjectResponse, status_code=201)
async def create_project(
    workspace_id: int,
    project_data: ProjectCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    创建项目
    
    在指定工作空间中创建一个新的项目。
    """
    project = await project_service.create_project(
        db=db,
        project_data=project_data,
        workspace_id=workspace_id,
        user_id=current_user.id
    )
    return await project_service.build_project_response(db, project)


@router.get("/", response_model=List[ProjectResponse])
async def list_projects(
    workspace_id: int,
    include_archived: bool = False,
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    列出工作空间的项目
    
    返回工作空间下的所有项目，支持分页。
    """
    projects = await project_service.list_workspace_projects(
        db=db,
        workspace_id=workspace_id,
        include_archived=include_archived,
        limit=limit,
        offset=offset
    )
    return [await project_service.build_project_response(db, p) for p in projects]


@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(
    project_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取项目详情
    
    返回项目的完整信息。
    """
    project = await project_service.get_project_by_id(db, project_id)
    return await project_service.build_project_response(db, project)


@router.patch("/{project_id}", response_model=ProjectResponse)
async def update_project(
    project_id: int,
    update_data: ProjectUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    更新项目
    
    可以更新项目的名称、描述、是否公开等属性。
    """
    project = await project_service.update_project(
        db=db,
        project_id=project_id,
        update_data=update_data
    )
    return await project_service.build_project_response(db, project)


@router.delete("/{project_id}", status_code=204)
async def delete_project(
    project_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    删除项目（软删除）
    
    项目不会真正删除，而是标记为已删除状态。
    """
    await project_service.delete_project(db, project_id)
    return None


# ==================== Project Archive ====================

@router.post("/{project_id}/archive", response_model=ProjectResponse)
async def archive_project(
    project_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    归档项目
    
    归档后的项目将从常规列表中隐藏。
    """
    project = await project_service.archive_project(db, project_id)
    return await project_service.build_project_response(db, project)


@router.post("/{project_id}/restore", response_model=ProjectResponse)
async def restore_project(
    project_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    恢复项目
    
    恢复已归档的项目。
    """
    project = await project_service.restore_project(db, project_id)
    return await project_service.build_project_response(db, project)


# ==================== Project Members ====================

@router.get("/{project_id}/members", response_model=List[dict])
async def list_project_members(
    project_id: int,
    only_active: bool = True,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    列出项目成员
    
    返回项目的所有成员及其角色。
    """
    members = await project_service.list_project_members(
        db=db,
        project_id=project_id,
        only_active=only_active
    )
    return [
        {
            "id": member.id,
            "user_id": member.user_id,
            "role": member.role,
            "is_active": member.is_active,
            "user": {
                "id": member.user.id,
                "display_name": member.user.display_name,
                "username": member.user.username,
                "avatar": member.user.avatar
            } if member.user else None
        }
        for member in members
    ]


@router.post("/{project_id}/members", response_model=dict)
async def add_project_member(
    project_id: int,
    user_id: int,
    role: int = Query(15, ge=1, le=20),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    添加项目成员
    
    将指定用户添加为项目成员。
    """
    member = await project_service.add_project_member(
        db=db,
        project_id=project_id,
        user_id=user_id,
        role=role,
        added_by=current_user.id
    )
    return {
        "id": member.id,
        "project_id": member.project_id,
        "user_id": member.user_id,
        "role": member.role
    }


@router.patch("/{project_id}/members/{user_id}", response_model=dict)
async def update_project_member(
    project_id: int,
    user_id: int,
    role: int = Query(..., ge=1, le=20),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    更新项目成员角色
    """
    member = await project_service.update_project_member(
        db=db,
        project_id=project_id,
        user_id=user_id,
        role=role,
        updated_by=current_user.id
    )
    return {
        "id": member.id,
        "project_id": member.project_id,
        "user_id": member.user_id,
        "role": member.role
    }


@router.delete("/{project_id}/members/{user_id}", response_model=dict)
async def remove_project_member(
    project_id: int,
    user_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    移除项目成员
    
    从项目中移除指定成员。
    """
    result = await project_service.remove_project_member(
        db=db,
        project_id=project_id,
        user_id=user_id
    )
    return result


# ==================== Project Statistics ====================

@router.get("/{project_id}/statistics", response_model=dict)
async def get_project_statistics(
    project_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取项目统计
    
    返回项目的工作项统计、进度、成员数等。
    """
    stats = await project_service.get_project_statistics(db, project_id)
    return stats


@router.get("/{project_id}/issues-summary", response_model=dict)
async def get_project_issues_summary(
    project_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取项目工作项摘要
    
    按状态分组返回工作项数量。
    """
    summary = await project_service.get_project_issues_summary(db, project_id)
    return summary